src/flb_file.c

```diff
@@ -0,0 +1,72 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2019-2021 The Fluent Bit Authors
+ *  Copyright (C) 2015-2018 Treasure Data Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_file.h>
+#include <fluent-bit/flb_log.h>
+#include <fluent-bit/flb_mem.h>
+#include <fluent-bit/flb_sds.h>
+
+#include <stdio.h>
+
+flb_sds_t flb_file_read(const char *path)
+{
+    long flen;
+    FILE *f = NULL;
+    flb_sds_t result = NULL;
+
+    f = fopen(path, "rb");
+    if (!f) {
+        return NULL;
+    }
+
+    if (fseek(f, 0, SEEK_END) == -1) {
+        goto err;
+    }
+
+    flen = ftell(f);
+    if (flen < 0) {
+        goto err;
+    }
+
+    if (fseek(f, 0, SEEK_SET) == -1) {
+        goto err;
+    }
+
+    result = flb_sds_create_size(flen);
+    if (!result) {
+        goto err;
+    }
+
+    if (flen > 0 && fread(result, flen, 1, f) != 1) {
```

Comment:
maybe here we need more tests to check the return value, per manpage:

```
If an error occurs, or the end of the file is reached, the return value is a short item count (or zero).
```

so there is a chance the return value is less than expect bytes to read

Reply:
This is the prototype of `size_t fread(void *ptr, size_t size, size_t nmemb, FILE *stream);`. Unlike the lower level `read` from unistd, `fread` is meant to read records/blocks from a file. The `size` parameter is the size of each "record",  `nmemb` is the "record count", and the return value is how many records were read.

If we were using `read` from `unistd.h` it would be possible for the read operation to be interrupted at a lower level and we would need to try again (`EAGAIN` error code). Usually the correct way to read using unistd/read is with a loop until `0` is returned (which means end of file).

But the way we use `fread` here is with  file size as the record size, so the only value it can return for success is `1`. That means if `1` is not returned, it means an unrecoverable failure for us.

Reply:
you are right, thanks for the info